### ISSUE TITLE
Few improvements to handling date_time on locations in algorithms

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#3789]
This is more a review of how date_time is currently handled for the different actions and some proposals for improvements.

So I went down the rabbit hole to understand the entire flow for date_time objects from input to path finding in preparation of #3765, and had some thoughts. Hope they make sense!

## For `isochrone` & `centroid`

In `Dijkstras::Expand` we could use the location-dependent timeinfos, so each input location's time is tracked for the appropriate branch. To me it looks like this should be pretty easy. So far, we compute all time_infos but only use the first one we find and that one then for all locations:

https://github.com/valhalla/valhalla/blob/7d05fc8f2e4bbefb5b55442a7d0d6dd85b25a52f/src/thor/dijkstras.cc#L366-L370

We could just index into the `time_infos` vector with the `pred.path_id()`. But that's until now only implemented in `centroid`, `isochrone` doesn't track edge labels per input location yet. It might not be a big problem to allow `isochrones` to do the same, and there it would be more useful. So far we **don't** allow `arrive_by` type for isochrones, but we could lift that since we have a reverse Dijkstra for that (and a undocumented `reverse` isochrone parameter). So it'd make sense to use it. But I admit I didn't follow through with investigating more. I kind of think it'd be the same behavior of the Dijkstra as currently, just that the edgelabels/status know about their path_id which is irrelevant as long as we don't use it (and still the costing will be more accurate in case we provide location-wise `date_time` strings). The initial adjacency list, after adding the origin/destination locations, will not change, neither will the algorithm itself, we'd just additionally look for the right time info for each location which will alter the weights somewhat when traffic info is present, but that's exactly what it should do.

For `centroid` I think it's best to just consider all location-dependent `date_time` objects, which is not much different from what is done now, only that we'd in the future allow a forward expansion from each location at a different time.. that's maybe mostly relevant if the location input to `centroid` really lives in different timezones, then they'd technically want to put different `date_time`s for each location. Actually I think I'd almost call it a bug: `Dijkstra::ExpandInner` will get the `offset_time` potentially based on wrong timezone info, the node might be in a different timezone than the `time_info` object it's diff'd to:

https://github.com/valhalla/valhalla/blob/master/src/thor/dijkstras.cc/#L145-L150

I also don't think it'd hurt the `reach` use of Dijkstra, as that doesn't pass time infos.

## Matrix

Depends on #3765, but I'll definitely very soon add time infos to `timedistancematrix`. For the forward Dijkstra (i.e. one:many) it'll represent the departure time of the origin(s). For the reverse Dijkstra in the many:one case it'll represent the arrival time at the destination(s). There's no other way basically. In either case, there can be one `date_time` for each source/target or overridden by the global `date_time.value`, while the global `date_time.type` is actually rather irrelevant though it shouldn't be `invariant` I guess, there we could throw.

## Route

I hardly looked into it tbh. It seems like `route` already respects each location's `date_time` string, so I'd imagine that even `through` locations get their own `departure` or `arrival` time, depending on if forward or reverse search was triggered. For me that's also a "aha" moment, wasn't really aware of that before.. But if people bother to set `date_time` for an intermediate waypoint, they should know what they're doing?

## Conclusion

- I'd first add time tracking to `timedistancematrix` as proposed above, that shouldn't be hard
- at a later point take care of the Dijkstra for `isochrone` & `centroid` if you agree on what I wrote

**Comment by nilsnolde** [comment]
@dnesbitt61 Sure, currently it's mostly used for ped/bicycles. We have a use case where we'd like to have matrix computations for motorized vehicles in a high-congestion city centre for e.g. food delivery where distances hardly ever go beyond 6-7 km. So the Dijkstra doesn't have much to do. We are aware of the performance penalties and could live with them for that particular (not so rare) use case.

Adding timeinfo is also pretty cheap (as long it's not actually being set in the request). They're only created once per location and then passed into `ExpandForward/Reverse` where they'll track the time (and bail early if there is no timeinfo set), and passing it into the various functions, where (I think) they'll actually don't do stuff you wouldn't want for bike/ped. And bike doesn't even use it in `EdgeCost`. Time info for bike/ped is even currently used in `master` for all other algorithms AFAICT (e.g. in `BycicleCost::Allowed` when deciding on time-dependent restrictions, in `'DynamicCost::Restricted` when deciding on time-dependent turn restrictions). So if someone happens to pass date_time info into a bicycle routing request to a server with traffic data, it looks like the unidirectional AStar and centroid/isochrone Dijkstra will do what I'd like the `TimeDistanceMatrix` to do. More as a side note, that's not really an argument.

Let me try to summarize based on what I think it'll end up impacting in `TimeDistanceMatrix`:
- `Allowed` will take the time info to pass it into `EvaluateRestrictions` where it will not take time-dependent restrictions if they kick in. Having a valid time info will actually disallow bike/ped on time restricted general restrictions (i.e. valid for all modes and I guess that includes bikes). I think that'd be even desired no? (even though it's questionable if that wouldn't be a tagging error, I've never seen bike-explicit time restrictions and I doubt, but don't know, if general time-dependent restrictions affect bikes/ped in any country?)
- Same for `DynamicCost::Restricted`
- `EdgeCost` is not affected for bike, and for ped only on ferry edges (which are few and it hardly matters IMO)

And any of this will only ever happen if there's actually a `date_time` object set or `date_time` string on any location, in which case I'd assume intention by the user and would actually want to respect time-dependent restrictions if they're supposedly valid for bikes (which might not even be a thing in OSM or in reality? was too lazy to query taginfo ;)). 

If that didn't convince, I can think of a few options how to handle it which would keep the current situation optionally:
- add a json config e.g. `allow_time_matrix` (stupid name, can think of smth better) as boolean which could just disable all time info in `thor.matrix_action::matrix`
- explicitly only disable timeinfo when the costing is bike/ped, e.g. also in matrix_action

Another thing popped in my head when walking the code: in either case, IMO it'd be nicer if we'd implicitly select the `TimeDistanceMatrix` algo when a request contains `date_time` stuff in either one of the locations or globally, and pass those time infos (or only if the config allows for it). With the current code, choosing `timedistancematrix` is _only_ a server-side config option for anything else than bike/ped/multimodal. If there's an additional protection in the form of a boolean config option for the server, I think that'd be a pretty safe thing to do, as then both, the server people and the requester want to allow for time-dependent routing in the matrix.

What do you think?

**Comment by dnesbitt61** [comment]
seems fine. We use one to many TimeDistanceMatrix in our embedded Android application for bicycle rerouting, so I'm a bit sensitive on performance. Seems like it won't impact much, but would be great to do some before and after performance checks. I expect using TimeDistanceMatrix for many to many driving matrix computations may be quite slow - though you mention this should be over very short distances so it could be ok. Perhaps the matrix_locations parameter could be useful (to shortcut the algorithm once a specified number of locations have been found) - I note this isn't in the documentation.

**Comment by nilsnolde** [comment]
> would be great to do some before and after performance checks

Jep, I'll run some `valhalla_run_matrix` with multi-run and report the results.

> Perhaps the matrix_locations parameter could be useful (to shortcut the algorithm once a specified number of locations have been found) - I note this isn't in the documentation.

Good point, I often forget about that one! It actually is in the documentation: https://github.com/valhalla/valhalla/blob/master/docs/api/matrix/api-reference.md#inputs-of-the-matrix-service. We could re-structure that a bit, to make it more obvious. In our case we want symmetrical matrices.

There's also bit of performance waiting over in [this issue](https://github.com/valhalla/valhalla/issues/3768), which I'd do in one go. Not that relevant for one:many's but (hopefully) quite relevant for 128x128 with pretty short distances in-between.

**Comment by nilsnolde** [comment]
And what do you think about this one @dnesbitt61 ? I think it'd be nice to have a boolean json config or even a setting allowing a number of `max_matrix_time_dependent_distance`, to have finer control over how much load someone wants to have on his server for those expensive things. If 0, we disable using timedistance for motorized vehicles and/or clear the time info if present.

> ... I can think of a few options how to handle it which would keep the current situation optional:
> 
> - add a json config e.g. allow_time_matrix (stupid name, can think of smth better) as boolean which could just disable all time info in thor.matrix_action::matrix
> - explicitly only disable timeinfo when the costing is bike/ped, e.g. also in matrix_action
